File: src/formats/srt.ts
// SubRip (.srt) adapter.
//
// Format gotchas:
//  - Time: "HH:MM:SS,mmm --> HH:MM:SS,mmm". Millisecond separator is ',' but we
//    tolerate '.' on input; we always emit ','.
//  - Cue numbers are regenerated on export (1-based), not trusted from input.
//  - Multi-line cue text: line breaks are PRESERVED as "\n" (unlike Lyrical Sync's
//    LRC→SRT path which joined lines with spaces). Subtitle text needs the breaks.
//  - Tokens (word timing) have no SRT representation → dropped on export.

import { newCueId, emptyDocument, type Cue, type SubtitleDocument } from "../types/subtitle";
import { formatSrtTime, parseClockTime } from "../utils/time";

const ARROW_RE = /-->/;

export function parseSrt(raw: string): SubtitleDocument {
  const doc = emptyDocument("srt");
  // Normalize newlines, split into blocks on blank lines.
  const blocks = raw.replace(/\r\n/g, "\n").replace(/^﻿/, "").split(/\n\s*\n/);

  for (const block of blocks) {
    const lines = block.split("\n");
    // Drop a leading numeric index line if present.
    if (lines.length && /^\d+$/.test(lines[0].trim())) lines.shift();
    if (!lines.length) continue;

    const timeLine = lines.shift()!;
    if (!ARROW_RE.test(timeLine)) continue;
    const [startStr, endStr] = timeLine.split(ARROW_RE);
    const start = parseClockTime(startStr.trim());
    const end = parseClockTime(endStr.trim());
    if (start == null || end == null) continue;

    const text = lines.join("\n").trim();
    doc.cues.push({ id: newCueId(), start, end, text });
  }
  return doc;
}

export function serializeSrt(doc: SubtitleDocument): string {
  const cues = sortedCues(doc.cues);
  const out = cues.map((cue, i) => {
    return `${i + 1}\n${formatSrtTime(cue.start)} --> ${formatSrtTime(cue.end)}\n${cue.text}`;
  });
  return out.join("\n\n") + "\n";
}

export function sortedCues(cues: Cue[]): Cue[] {
  return [...cues].sort((a, b) => a.start - b.start || a.end - b.end);
}

File: native/Sources/GlyphlineCore/Formats/SRT.swift
// SubRip (.srt) adapter (ported from ../../src/formats/srt.ts).
//
//  - Time: "HH:MM:SS,mmm --> HH:MM:SS,mmm" (tolerate '.' on input, emit ',').
//  - Leading numeric index line dropped on parse, regenerated (1-based) on export.
//  - Multi-line text preserved as "\n". Tokens have no SRT form → dropped on export.

import Foundation

public func parseSrt(_ raw: String) -> SubtitleDocument {
    var doc = SubtitleDocument.empty(.srt)
    let normalized = raw.replacingOccurrences(of: "\r\n", with: "\n").strippingLeadingBOM()
    for block in regexSplit(#"\n\s*\n"#, normalized) {
        var lines = block.components(separatedBy: "\n")
        // Drop a leading numeric index line if present.
        if let first = lines.first, regexTest(#"^\d+$"#, first.trimmed()) { lines.removeFirst() }
        guard !lines.isEmpty else { continue }
        let timeLine = lines.removeFirst()
        guard timeLine.contains("-->") else { continue }
        let parts = timeLine.components(separatedBy: "-->")
        guard parts.count == 2,
              let start = parseClockTime(parts[0].trimmed()),
              let end = parseClockTime(parts[1].trimmed()) else { continue }
        let text = lines.joined(separator: "\n").trimmed()
        doc.cues.append(Cue(id: newCueId(), start: start, end: end, text: text))
    }
    return doc
}

public func serializeSrt(_ doc: SubtitleDocument) -> String {
    let cues = sortedCues(doc.cues)
    let blocks = cues.enumerated().map { i, cue in
        "\(i + 1)\n\(formatSrtTime(cue.start)) --> \(formatSrtTime(cue.end))\n\(cue.text)"
    }
    return blocks.joined(separator: "\n\n") + "\n"
}
